filters/valve.cpp

```diff
@@ -0,0 +1,205 @@
+  }
+
+  Future<bool> isOpened() {
+    return this->opened;
+  }
+
+  const lambda::function<Future<bool>()> getIsOpenedFunction() {
+    return defer(self(), &Self::isOpened);
+  }
+
+ protected:
+  virtual void initialize() {
+    switch (valveType) {
+      case RESOURCE_ESTIMATOR_VALVE:
+        route(VALVE_ROUTE,
+              ESTIMATOR_VALVE_ENDPOINT_HELP(),
+              &ValveFilterEndpointProcess::valve);
+        LOG(INFO) << "Serenity Resource Estimator endpoint initialized "
+                  << "on /" << RESOURCE_ESTIMATOR_PROCESS_BASE  << VALVE_ROUTE;
+        break;
+      case QOS_CONTROLLER_VALVE:
+        route(VALVE_ROUTE,
+              CONTROLLER_VALVE_ENDPOINT_HELP(),
+              &ValveFilterEndpointProcess::valve);
+        LOG(INFO) << "Serenity QoS Controller endpoint initialized "
+                  << "on /" << QOS_CONTROLLER_PROCESS_BASE  << VALVE_ROUTE;
+        break;
+    }
+  }
+
+ private:
+  Future<http::Response> valve(const http::Request& request) {
+    return limiter.acquire()
+      .then(defer(self(), &Self::_valve, request));
+  }
+
+  Future<http::Response> _valve(const http::Request& request) {
+    Try<hashmap<string, string>> decode =
+        process::http::query::decode(request.body);
+    if (decode.isError()) {
+      return http::BadRequest(
+          "Unable to decode query string: " + decode.error());
+    }
+    hashmap<string, string> values = decode.get();
+
+    // Get params.
+    string enabled_param;
+    Option<string> pipeline_enable = request.query.get(PIPELINE_ENABLE_KEY);
+    if (pipeline_enable.isSome()) {
+      enabled_param = pipeline_enable.get();
+    } else {
+      Try<string> pipeline_enable_form =
+          getFormValue(PIPELINE_ENABLE_KEY, values);
+
+      if (pipeline_enable_form.isError()) {
+        return http::BadRequest(pipeline_enable_form.error());
+      }
+      enabled_param = pipeline_enable_form.get();
+    }
+
+    bool pipeline_enable_decision = false;
+    if (!enabled_param.compare("true")) {
+      pipeline_enable_decision = true;
+    }
+
+    this->setOpen(pipeline_enable_decision);
+
+    JSON::Object response;
+    response.values[PIPELINE_ENABLE_KEY] = enabled_param;
+
+    return http::OK(response);
+  }
+
+  atomic_bool opened;
+  ValveType valveType;
+  //! Used to rate limit the endpoint.
+  RateLimiter limiter;
+};
+
+
+ValveFilter::ValveFilter(
+    Consumer<ResourceUsage>* _consumer,
+    ValveType _valveType,
+    bool _opened)
+  : Producer<ResourceUsage>(_consumer),
+    process(new ValveFilterEndpointProcess(_valveType, _opened)) {
+  isOpened = process.get()->getIsOpenedFunction();
+  spawn(process.get());
+}
+
+
+ValveFilter::~ValveFilter() {
+  terminate(process.get());
+  wait(process.get());
+}
+
+
+Try<Nothing> ValveFilter::consume(const ResourceUsage& in) {
+  if (this->isOpened().get()) {
+    this->produce(in);
+  }
+
+  // TODO(bplotka): Should we continue the pipeline without any Executor in
+  // usage?
```

Comment:
Continuing pipeline = run *produce()* and pass some form of ResourceUsage further. Here i am just __aborting__ pipeline by just returning Nothing here. (: 

We should desing failover behaviour to be consistent in every filter.. Should we stop pipeline, when need for oversubscription cut off occured? Should we pass empty ResourceUsage? What if some executor does not have required fields? Should we check for that in every filter? Maybe one validation filter at the front of pipeline will be enough?  We don't have consistent view of that. I will gather these questions and start a discussion with you guys, since it's sth which should be defined. (: